### PR TITLE
DR-287 snapshot filesystem

### PR DESCRIPTION
**Pull request by ddietterich**
This adds the snapshot file system to the DR-515 work.
Once DR-515 goes in, I'll rebase this to develop and it can get reviewed.
DR-555 is the last part of the work.

### CODE REVIEW

File: src/main/java/bio/terra/filesystem/FireStoreDao.java [hunk 1]
```diff
@@ -92,6 +95,41 @@ public FireStoreFile lookupFile(Dataset dataset, String objectId) {
         return fileDao.retrieveFileMetadata(firestore, datasetId, objectId);
     }
 
+    public void addFilesToSnapshot(Dataset dataset, Snapshot snapshot, List<String> refIds) {
+        Firestore datasetFirestore = FireStoreProject.get(dataset.getDataProjectId()).getFirestore();
+        Firestore snapshotFirestore = FireStoreProject.get(snapshot.getDataProjectId()).getFirestore();
+        String datasetId = dataset.getId().toString();
+        // TODO: Do we need to clean up the database name?
```
Comment (thathert) [inline src/main/java/bio/terra/filesystem/FireStoreDao.java:102]: I'm confused by this line..is this referring to the dataset name?

File: src/main/java/bio/terra/filesystem/FireStoreDirectoryDao.java [hunk 2]
```diff
@@ -347,4 +312,120 @@ private String makeLookupPath(String fullPath) {
     private String makePathFromLookupPath(String lookupPath) {
         return StringUtils.removeStart(lookupPath, ROOT_DIR_NAME);
     }
+
+    // -- Snapshot filesystem methods --
+
+    /**
+     * Given ane object id from a dataset directory, create a similar object in the snapshot directory.
```
Comment (thathert) [inline src/main/java/bio/terra/filesystem/FireStoreDirectoryDao.java:319]: typo at `ane`

File: src/main/java/bio/terra/filesystem/FireStoreDirectoryDao.java [hunk 3]
```diff
@@ -347,4 +312,120 @@ private String makeLookupPath(String fullPath) {
...
+        }
+
+        FireStoreObject topDir = new FireStoreObject()
+            .objectId(UUID.randomUUID().toString())
+            .fileRef(false)
+            .path("/")
+            .name(dirName)
+            .fileCreatedDate(Instant.now().toString());
+
+        createFileRef(firestore, collectionId, topDir);
```
Comment (thathert) [inline src/main/java/bio/terra/filesystem/FireStoreDirectoryDao.java:381]: This name feels off -- we are passing `topDir`, which is not a fileRef, to `createFileRef`. Maybe a more accurate name is something along the lines of `createObjectAndParents`?

File: src/main/java/bio/terra/flight/snapshot/create/CreateSnapshotFireStoreDataStep.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+        // Build the snapshot file system and record the file dependencies
+        // The algorithm is:
+        // Loop through sources, loop through map tables, loop through map columns
+        // if from column is FILEREF or DIRREF, ask pdao to get the ids from that
+        // column that are in the snapshot; do the filestore processing
+        //
+        // NOTE: This is brute force doing a column at a time. Depending on how much memory and
+        // swap we want to use, we could extract all row ids in one go. Doing it column-by-column
+        // bounds the intermediate size in a way. I think this all becomes easier if we move
```
Comment (thathert) [inline src/main/java/bio/terra/flight/snapshot/create/CreateSnapshotFireStoreDataStep.java:57]: The end of this comment is a little out of date :)

**Review by thathert — APPROVED** [review]
LGTM

File: src/main/java/bio/terra/filesystem/FireStoreDao.java (hunk 1 above)
Comment (ddietterich) [reply src/main/java/bio/terra/filesystem/FireStoreDao.java:102]: Yes, I meant dataset name. The question is: could dataset name have characters that are not allowed in filesystem named (like / and 0x1c).

I made the comment more specific in DR-555 and added the work to understand if anything needs to be done to DR-325.

File: src/main/java/bio/terra/filesystem/FireStoreDirectoryDao.java (hunk 2 above)
Comment (ddietterich) [reply src/main/java/bio/terra/filesystem/FireStoreDirectoryDao.java:319]: Its Olde English 😁

File: src/main/java/bio/terra/filesystem/FireStoreDirectoryDao.java (hunk 3 above)
Comment (ddietterich) [reply src/main/java/bio/terra/filesystem/FireStoreDirectoryDao.java:381]: Yes, that is kind of hopeless. I'll add it to my grand rename.

File: src/main/java/bio/terra/flight/snapshot/create/CreateSnapshotFireStoreDataStep.java (hunk 4 above)
Comment (ddietterich) [reply src/main/java/bio/terra/flight/snapshot/create/CreateSnapshotFireStoreDataStep.java:57]: Indeed. Removed in DR-555

**Review by RoriCremer — APPROVED** [review]

**Pull request merged**